File: backend/internal/domain/errors/domain_error.go
package errors

type Code string

const (
	CodeValidationError   Code = "VALIDATION_ERROR"
	CodeNotFound          Code = "NOT_FOUND"
	CodeDuplicate         Code = "DUPLICATE"
	CodeUnauthorized      Code = "UNAUTHORIZED"
	CodeForbidden         Code = "FORBIDDEN"
	CodeInvalidState      Code = "INVALID_STATE"
	CodeInvalidTransition Code = "INVALID_TRANSITION"
	CodeConflict          Code = "CONFLICT"
)

type DomainError struct {
	Code    Code
	Message string
	Details map[string]any
}

func (e *DomainError) Error() string {
	return e.Message
}

func New(code Code, message string, details map[string]any) error {
	return &DomainError{
		Code:    code,
		Message: message,
		Details: details,
	}
}

File: frontend/src/dominio/errores/DomainError.ts
export type DomainErrorCode =
  | 'VALIDATION_ERROR'
  | 'NOT_FOUND'
  | 'DUPLICATE'
  | 'UNAUTHORIZED'
  | 'FORBIDDEN'
  | 'INVALID_STATE'
  | 'INVALID_TRANSITION'
  | 'CONFLICT'

export class DomainError extends Error {
  readonly code: DomainErrorCode
  readonly details?: Record<string, unknown>

  constructor(
    code: DomainErrorCode,
    message: string,
    details?: Record<string, unknown>,
  ) {
    super(message)
    this.name = 'DomainError'
    this.code = code
    this.details = details
  }
}

export const domainError = (
  code: DomainErrorCode,
  message: string,
  details?: Record<string, unknown>,
) => new DomainError(code, message, details)
